File: LeetCode/JavaScript/#125 Valid Palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
// First
var isPalindrome = function(s) {
    if(s==="") return true;
    s = s.replace(/\W+/g,"").toLowerCase();
    return s==s.split("").reverse().join("");
    var len = s.length;
    if(len%2===0){
      return s.slice(0,len/2)===s.slice(len/2).split("").reverse().join("");
    }else{
      len = Math.floor(len/2);
      return s.slice(0,len)===s.slice(len+1).split("").reverse().join("");
    }
};

// Second
var isPalindrome = function(s) {
    if(s==="") return true;
    s = s.replace(/\W+/g,"").toLowerCase();
    return s==s.split("").reverse().join("");
}


// others way faster than mine
var isPalindrome = function(s) {
    var str = s.replace(/\W/g, '').toLowerCase();
    var left = 0;
    var right = str.length - 1;
    
    while(left < right){
        if(str.charAt(left) !== str.charAt(right)) return false;
        left++;
        right--;
    }
    return true;
};

File: LeetCode/Python/#125 Valid Palindrome.py
import re
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if s=="":return True
        s = re.sub(r"\W+","",s).lower()
        left = 0
        right = len(s)-1

        while left < right:
            if s[left]!=s[right]:
                return False
            left +=1
            right -=1

        return True

# an easy way
def isPalindrome(self, s):
    p = re.compile("[\W_]+")
    s = p.sub("",s.lower())
    return s == s[::-1]

import re
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if s=="":return True
        s = re.sub(r"\W+","",s).lower()
        return s==s[::-1]